### PR TITLE
Integrate Autonomy with RoveComm and SIM

### PR DESCRIPTION
**Pull request by ClayJay3**
# Main Changes:
- **RoveComm**:
    - Update to new RoveComm commit.
    - Added cross-language support
    - Fixed double and float corruption.
    - Manifest updates
- **Added all RoveComm callbacks to Autonomy_Software code.**
    - NavBoard (GPS, Compass, Accuracy)
    - DriveBoard (SendDrive, Stop, SetMaxSpeed)
    - MultimediaBoard (Lighting RGB and LightingState)
    - StateMachineHandler (Autonomy Start/Stop)
    - WaypointHandler (Add Nav/Marker/Object waypoints and ClearWaypoints)
- **Changed the devcontainer network type to share the host machines network stack**.
    - This allows the use of the host machines 127.0.0.1 address to work with Autonomy_Software when sending packets with RoveCommTester
- **Logging**:
    - Added logging handler filter function so that the console and file handlers can have different logging levels. Currently the file handler is set to log everything and the console logger is set to output anything with a logging level greater than or equal to INFO.
- **StateMachine**
    - I have added MultimediaBoard commands to the TriggerEvent method for all states.
    - Made saving states on TriggerEvent() toggle-able, false by default.
    - General cleanup.
    - Change log statements for TriggerEvent to INFO.
    - **I have implemented the basic GPS navigation waypoint logic for the state machine, we can now navigate to queued GPS waypoints autonomously**

## Other Changes
- Tested drive to heading functionality with PID controller and inverse kinematics: everything works.
    - Fixed bug with Arcade and Curvature drive methods to make turning smoother and more consistent.
    - Fixed range mapping for setting max drive output power.
- Autonomy_Simulator and RoveComm_Tester_Software repos have undergone lots of changes. If you want to test in the SIM, make sure to pull/sync both repos.

### CODE REVIEW

File: src/AutonomyConstants.h [hunk 1]
```diff
@@ -30,20 +30,18 @@
...
-    //// RoveComm Constants.
-    ///////////////////////////////////////////////////////////////////////////
+    // Logging constants.
+    const std::string LOGGING_OUTPUT_PATH_ABSOLUTE = "./logs/";    // The absolute to write output logging and video files to.
 
-    // Socket Ports.
-    const int ROVECOMM_UDP_PORT                 = 11000;    // The UDP socket port to use for the main UDP RoveComm instance.
-    const int ROVECOMM_TCP_PORT                 = 12000;    // The UDP socket port to use for the main UDP RoveComm instance.
+    // RoveComm constants.
+    const int ROVECOMM_OUTGOING_UDP_PORT        = 11001;    // The UDP socket port to use for the main UDP RoveComm instance.
```
Comment (Byrdman32) [inline src/AutonomyConstants.h]: Thoughts on having this be set conditionally based upon the `MODE_SIM` constant?

File: src/AutonomyLogging.cpp [hunk 2]
```diff
@@ -106,17 +106,21 @@ namespace logging
         // Start Quill
         quill::start();
 
+        // Set Handler Filters
+        qFileHandler->add_filter(std::make_unique<LoggingFilter>("FileFilter", quill::LogLevel::TraceL3));
+        qConsoleHandler->add_filter(std::make_unique<LoggingFilter>("StdoutFilter", quill::LogLevel::Info));
```
Comment (Byrdman32) [inline src/AutonomyLogging.cpp]: Maybe name the filter `ConsoleFilter` instead to fit with the name of the logger it's being attached to?

File: src/AutonomyLogging.h [hunk 3]
```diff
@@ -63,5 +63,64 @@ namespace logging
...
+                                        std::chrono::nanoseconds log_message_timestamp,
+                                        quill::MacroMetadata const& metadata,
+                                        quill::fmt_buffer_t const& formatted_record) noexcept override
+            {
+                // Not using these.
+                (void) thread_id;
+                (void) log_message_timestamp;
+                (void) formatted_record;
+
+                // Log only WARNINGS or higher to stdout.
```
Comment (Byrdman32) [inline src/AutonomyLogging.h]: Since `eMinLogLevel` could technically not be set to `WARNING` this comment should be changed.

File: src/main.cpp [hunk 4]
```diff
@@ -88,21 +88,36 @@ int main()
...
         // Setup global objects.
         /////////////////////////////////////////
         // Initialize RoveComm.
         globals::g_pRoveCommUDPNode = new rovecomm::RoveCommUDP();
         globals::g_pRoveCommTCPNode = new rovecomm::RoveCommTCP();
         // Start RoveComm instances bound on ports.
-        bool bRoveCommUDPInitSuccess = globals::g_pRoveCommUDPNode->InitUDPSocket(constants::ROVECOMM_UDP_PORT);
-        bool bRoveCommTCPInitSuccess = globals::g_pRoveCommTCPNode->InitTCPSocket(constants::ROVECOMM_TCP_INTERFACE_IP.c_str(), constants::ROVECOMM_TCP_PORT);
+        bool bRoveCommUDPInitSuccess = globals::g_pRoveCommUDPNode->InitUDPSocket(manifest::General::ETHERNET_UDP_PORT);
+        bool bRoveCommTCPInitSuccess = globals::g_pRoveCommTCPNode->InitTCPSocket(constants::ROVECOMM_TCP_INTERFACE_IP.c_str(), manifest::General::ETHERNET_UDP_PORT);
```
Comment (Byrdman32) [inline src/main.cpp]: Should be `manifest::General::ETHERNET_TCP_PORT`???

**Review by Byrdman32 — CHANGES_REQUESTED** [review]

File: src/AutonomyConstants.h (hunk 1 above)
Comment (ClayJay3) [reply src/AutonomyConstants.h]: I agree, can you go ahead and make those changes?

Comment (Byrdman32) [reply src/AutonomyConstants.h]: Do you also want TCP changed or not?

Comment (ClayJay3) [reply src/AutonomyConstants.h]: Yeah, might as well.

**Review by Byrdman32 — APPROVED** [review]

**Pull request merged**